include/deal.II/matrix_free/matrix_free.templates.h

```diff
@@ -442,6 +442,26 @@ MatrixFree<dim, Number, VectorizedArrayType>::internal_reinit(
   // determined in @p extract_local_to_global_indices.
   if (additional_data.initialize_mapping == true)
     {
+      if (dof_handler.size() > 1)
+        {
+          // check if all DoHandlers are in the same hp mode; and if hp
+          // capabilities are enabled: check if active_fe_indices of all
+          // DoFHandler are the same.
+          for (unsigned int i = 1; i < dof_handler.size(); ++i)
+            {
+              Assert(dof_handler[0]->has_hp_capabilities() ==
+                       dof_handler[i]->has_hp_capabilities(),
+                     ExcNotImplemented());
```

Comment:
Do we plan to also allow the case of combining `hp` with non-`hp` cases? In a multiphysics context it could make sense to have high polynomial degrees in one component but not in the other. But one might also argue that the slight increase in efficiency from allowing those in the same `MatrixFree` is only a step towards allowing completely different meshes and the new evaluator from #11046, where an extension like this is unnecessary.

Reply:
I don't consider these asserts as permanent. In particular, would like to lift `dof_info[0].cell_active_fe_index == dof_info[i].cell_active_fe_index`. But these asserts represent the current state within `MatrixFree` and such are useful.